### ISSUE TITLE
The motivation of this work

### ISSUE THREAD

**Issue by SY-Xuan** [ExplainableML/LanguageGuidance_for_DML#6]
Thanks for your good work! I have some questions about your work.
The motivation of your work is to use the language model to guide the training of the visual model. Therefore, you use the language model to generate the soft target for the visual model. 
I think that the target and prediction of Eq (2) should be Inverted.
![](https://latex.codecogs.com/svg.image?\mathcal{L}_{\text&space;{match&space;}}\left(\mathcal{S}^{\text&space;{img&space;}},&space;\mathcal{S}^{\text&space;{lang&space;}}\right)=$$\quad&space;\frac{1}{|\mathcal{B}|}&space;\sum_{i}^{|B|}&space;\sigma\left(S_{i}^{\text&space;{lang&space;},&space;X}\right)&space;\log&space;\left(\frac{\sigma\left(S_{i}^{\text&space;{lang&space;},&space;X}\right)}{\sigma\left(S_{i}^{\text&space;{img&space;}}&plus;\gamma_{\text&space;{img&space;}}\right)}\right))

This corresponds to your implementation of **forward** distillation. However, you use the **backward** distillation.
The **backward** distillation is inconsistent with your motivation. 
![](https://latex.codecogs.com/svg.image?\mathcal{L}_{\text&space;{match&space;}}\left(\mathcal{S}^{\text&space;{img&space;}},&space;\mathcal{S}^{\text&space;{lang&space;}}\right)=$$\quad&space;\frac{1}{|\mathcal{B}|}&space;\sum_{i}^{|B|}&space;\sigma\left(S_{i}^{\text&space;{img&space;},&space;X}\right)&space;\log&space;\left(\sigma\left(S_{i}^{\text&space;{img&space;},&space;X}\right)\right)&space;-&space;\sigma\left(S_{i}^{\text&space;{img&space;},&space;X}\right)\log\left({\sigma\left(S_{i}^{\text&space;{lang&space;}}&plus;\gamma_{\text&space;{lang&space;}}\right)}\right))
The first term will increase the entropy of the visual model, leading to a uniform distribution. The second term leads to the one-hot prediction. If writing like this, the gradients of the first term cannot be ignored. However, I think the behavior of this Eq (2) is unpredictable.

**Comment by Confusezius** [comment]
Appreciate the kind words, and thanks for pointing this out!
I will make sure to add this information in the README! 

For now, in practice it only makes a minor difference swapping between either direction, as the crossentropy between language similarity distribution and visual similarity distribution is minimized (since `KLD = CE(LANG, VIS) - H(LANG OR VIS, depends on direction)`). In the backward case, the introduction of a differentiable `H(VIS)`-term has some nice regulatory properties, as entropy maximization during training is known to benefit generalisation. I wouldn't go so far as calling the behaviour "unpredictable", as it simply introduces an additional entropy regularizer during training :). 
And this only adds minimal benefits, as switching to forward distillation provides nearly similar improvements (the direction was treated as a hyperparameter during grid-searching, with minimal benefits which is why it was chosen as default here :)).

**Comment by SY-Xuan** [comment]
Thanks for your kindly replying. If we ignore H(VIS), CE(LANG, VIS) and CE(VIS, LANG) are still different(The gradients are different). I still cannot understand the effect of performing backward distillation. However, the forward distillation is straightforward.
